geomesa-stream/geomesa-stream-datastore/src/main/scala/org/locationtech/geomesa/stream/datastore/StreamDataStore.scala

```diff
@@ -0,0 +1,214 @@
+package org.locationtech.geomesa.stream.datastore
+
+import java.awt.RenderingHints
+import java.util.Collections
+import java.util.concurrent.{Executors, TimeUnit}
+import java.util.logging.Level
+import java.{util => ju}
+
+import com.google.common.cache.{Cache, CacheBuilder, RemovalListener, RemovalNotification}
+import com.google.common.collect.Lists
+import com.typesafe.config.ConfigFactory
+import com.vividsolutions.jts.geom.{Envelope, Geometry}
+import org.geotools.data.DataAccessFactory.Param
+import org.geotools.data._
+import org.geotools.data.collection.DelegateFeatureReader
+import org.geotools.data.store.{ContentDataStore, ContentEntry, ContentFeatureSource, ContentFeatureStore}
+import org.geotools.factory.CommonFactoryFinder
+import org.geotools.feature.collection.DelegateFeatureIterator
+import org.geotools.filter.FidFilterImpl
+import org.geotools.geometry.jts.{JTS, ReferencedEnvelope}
+import org.geotools.referencing.crs.DefaultGeographicCRS
+import org.locationtech.geomesa.stream.SimpleFeatureStreamSource
+import org.locationtech.geomesa.utils.geotools.Conversions._
+import org.locationtech.geomesa.utils.index.SynchronizedQuadtree
+import org.opengis.feature.`type`.Name
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.expression.{Literal, PropertyName}
+import org.opengis.filter.spatial.{BBOX, BinarySpatialOperator, Within}
+import org.opengis.filter.{And, Filter, IncludeFilter, Or}
+
+import scala.collection.JavaConversions._
+
+case class FeatureHolder(sf: SimpleFeature, env: Envelope) {
+  override def hashCode(): Int = sf.hashCode()
+
+  override def equals(obj: scala.Any): Boolean = obj match {
+    case other: FeatureHolder => sf.equals(other.sf)
+    case _ => false
+  }
+}
+
+trait StreamListener {
+  def onNext(sf: SimpleFeature): Unit
+}
+
+class StreamDataStore(source: SimpleFeatureStreamSource, timeout: Int) extends ContentDataStore {
+  
+  val sft = source.sft
+  source.init()
+  val qt = new SynchronizedQuadtree
+
+  val cb =
+    CacheBuilder
+      .newBuilder()
+      .expireAfterWrite(timeout, TimeUnit.SECONDS)
+      .removalListener(
+        new RemovalListener[String, FeatureHolder] {
+          def onRemoval(removal: RemovalNotification[String, FeatureHolder]) = {
+            qt.remove(removal.getValue.env, removal.getValue.sf)
+          }
+        }
+      )
+
+  val features: Cache[String, FeatureHolder] = cb.build()
+
+  val listeners = Collections.synchronizedList(Lists.newArrayList[StreamListener]())
```

Comment:
since this is presumably almost always read and very rarely write, you could use a concurrent collection instead:
http://docs.oracle.com/javase/6/docs/api/java/util/concurrent/CopyOnWriteArrayList.html